Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  - type: Reconciled
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  shardStatuses:
+  - shardID: "0"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 1
+    unavailableReplicas: 1
+  - shardID: "1"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 1
+    unavailableReplicas: 1
+```
+
+### Alertmanager
+
+The subresource status for the Alertmanager custom resource definition should
+be very similar to the API defined for the Prometheus CRD. The main difference
+is that we don't need the `shardStatuses` field.
+
+### Thanos Ruler
+
+The subresource status for the Thanos Ruler custom resource definition is
+identical to the Alertmanager CRD.
+
+### Service Monitor
+
+```golang
+type ServiceMonitor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of desired Service selection for target discovery by
+	// Prometheus.
+	Spec ServiceMonitorSpec `json:"spec"`
+	// 
+	// Most recent observed status of the service monitor.
+	// Populated by the Prometheus operator.
+	// Read-only.
+	Status ServiceMonitorStatus `json:"status"`
+}
+
+type ServiceMonitorStatus struct {
+	// The current state of the service monitor.
+	// +patchMergeKey=type
+	// +patchMergeStrategy=merge
+	// +optional
+	Conditions []ServiceMonitorStatusCondition `json:"conditions,omitempty"`
+
+	// The list of resources that the service monitor is bound to.
+	// +patchMergeKey=name
+	// +patchMergeStrategy=merge
+	// +optional
+	Bindings []ServiceMonitorBindings `json:"bindings,omitempty"`
+}
+
+type ServiceMonitorStatusCondition struct {
+	// Type of the condition being reported.
+	// +required
+	Type ServiceMonitorConditionType `json:"type"`
+	// status of the condition.
+	// +required
+	Type ServiceMonitorConditionStatus `json:"status"`
+	// lastTransitionTime is the time of the last update to the current status property.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	// Reason for the condition's last transition.
+	// +optional
+	Reason *string `json:"reason,omitempty"`
+	// Human-readable message indicating details for the condition's last transition.
+	// +optional
+	Message *string `json:"reason,omitempty"`
+}
+
+type ServiceMonitorConditionType string
+
+const (
+	// Valid indicates if the operator considers the service monitor to be valid.
+	ServiceMonitorValid ServiceMonitorConditionType = "Valid"
+	// Selected indicates if the service monitor is reconciled by the referenced Prometheus object.
+	ServiceMonitorReconciled ServiceMonitorConditionType = "Reconciled"
+)
+
+type ServiceMonitorConditionStatus string
+
+const (
+	ServiceMonitorConditionTrue    ServiceMonitorConditionStatus = "True"
+	ServiceMonitorConditionFalse   ServiceMonitorConditionStatus = "False"
+	ServiceMonitorConditionUnknown ServiceMonitorConditionStatus = "Unknown"
+)
+
+type ServiceMonitorBinding struct {
```

Comment:
Would it make sense to use upstream `[(k8s.io/api/core/v1).Binding](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5339-L5348) instead?

Reply:
This type is marked as deprecated. In general, I got the impression that it's better to declare custom types than relying on core types (for instance [v1.ObjectReference](https://github.com/kubernetes/api/blob/4097bdc1441167140bb232b0246dcc1e3b8f40e5/core/v1/types.go#L5556-L5557)).